src/dac.rs

```diff
@@ -0,0 +1,279 @@
+    next_buffer: Option<&'static mut [u16; SAMPLE_BUFFER_SIZE]>,
+    _spi: hal::spi::Spi<hal::stm32::SPI4, hal::spi::Disabled, u16>,
+    transfer: Transfer<
+        hal::dma::dma::Stream4<hal::stm32::DMA1>,
+        SPI4,
+        MemoryToPeripheral,
+        &'static mut [u16; SAMPLE_BUFFER_SIZE],
+    >,
+    first_transfer: bool,
+}
+
+impl Dac0Output {
+    /// Construct the DAC0 output channel.
+    ///
+    /// # Args
+    /// * `spi` - The SPI interface used to communicate with the ADC.
+    /// * `stream` - The DMA stream used to write DAC codes over SPI.
+    /// * `trigger_channel` - The sampling timer output compare channel for update triggers.
+    pub fn new(
+        spi: hal::spi::Spi<hal::stm32::SPI4, hal::spi::Enabled, u16>,
+        stream: hal::dma::dma::Stream4<hal::stm32::DMA1>,
+        trigger_channel: sampling_timer::Timer2Channel3,
+    ) -> Self {
+        // Generate DMA events when an output compare of the timer hitting zero (timer roll over)
+        // occurs.
+        trigger_channel.listen_dma();
+        trigger_channel.to_output_compare(0);
+
+        // The stream constantly writes to the TX FIFO to write new update codes.
+        let trigger_config = DmaConfig::default()
+            .memory_increment(true)
+            .peripheral_increment(false);
+
+        // Construct the trigger stream to write from memory to the peripheral.
+        let transfer: Transfer<_, _, MemoryToPeripheral, _> = Transfer::init(
+            stream,
+            SPI4::new(),
+            unsafe { &mut DAC0_BUF0 },
+            None,
+            trigger_config,
+        );
+
+        // Listen for any potential SPI error signals, which may indicate that we are not generating
+        // update codes.
+        let mut spi = spi.disable();
+        spi.listen(hal::spi::Event::Error);
+
+        // Allow the SPI FIFOs to operate using only DMA data channels.
+        spi.enable_dma_tx();
+
+        // Enable SPI and start it in infinite transaction mode.
+        spi.inner().cr1.modify(|_, w| w.spe().set_bit());
+        spi.inner().cr1.modify(|_, w| w.cstart().started());
+
+        Self {
+            transfer,
+            next_buffer: unsafe { Some(&mut DAC0_BUF1) },
+            _spi: spi,
+            first_transfer: true,
+        }
+    }
+
+    /// Schedule the next set of DAC update codes.
+    ///
+    /// # Args
+    /// * `data` - The next samples to enqueue for transmission.
+    pub fn next_data(&mut self, data: &[u16; SAMPLE_BUFFER_SIZE]) {
+        let next_buffer = self.next_buffer.take().unwrap();
+
+        // Copy data into the next buffer
+        next_buffer.copy_from_slice(data);
+
+        // If the last transfer was not complete, we didn't write all our previous DAC codes.
+        // Wait for all the DAC codes to get written as well.
+        if self.first_transfer {
+            self.first_transfer = false
+        } else {
+            while self.transfer.get_transfer_complete_flag() == false {}
```

Comment:
Do we want to stall here? What if we need to need to service the other DAC?

Reply:
Stalling here is the simpler solution from a code standpoint. We won't be stalling very long

Reply:
Added comment + priority change to these loops to indicate they are not checked formaximum times for speed purposes.